text/trace/0157-messaging-semantic-conventions.md

```diff
@@ -0,0 +1,238 @@
+settle batches of messages. However, it is not possible to settle messages
+independent of their position in the stream (e. g., if message B is located at
+a later position in the stream than message A, then message B cannot be settled
+without also settling message A).
+
+```
+                               Checkpoint
+                                   | 
+                                   v                                
++---------+ +---------+ +---------+ +---------+ +---------+ +---------+
+|Message A| |Message B| |Message C| |Message D| |Message E| |Message F|
++---------+ +---------+ +---------+ +---------+ +---------+ +---------+
+                     <---  Settled
+```
+
+#### Examples
+
+1. The following configurations should be instrumented and tested for Kafka or
+   a similar messaging system:
+
+   * 1 producer, 2 consumers in the same consumer group
+   * 1 producer, 2 consumers in different consumer groups
+   * 2 producers, 2 consumers in the same consumer group
+
+   Each of the producers produces a continuous stream of messages.
+
+## Open questions
+
+The following areas are considered out-of-scope of a first stable release of
+semantic conventions for messaging. While not being explicitly considered for
+a first stable release, it is important to ensure that this first stable
+release can serve as a solid foundation for further improvements in these areas.
+
+### Sampling
+
+The current experimental semantic conventions rely heavily on span links as
+a way to correlate spans. This is necessary, as several traces are needed to
+model the complete path that a message takes through the system. With the currently
+available sampling capabilities of OpenTelemetry, it is not possible to ensure
+that a set of linked traces is sampled. As a result, it is unlikely to sample a
+set of traces that covers the complete path a message takes.
+
+Solving this problem requires a solution for sampling based on span links,
+which is not in scope for this OTEP.
+
+However, having a too high number of span links in a single trace or having too
+many traces linked together can make the visualization and analysis of traces
+inefficient. This problem is not related to sampling and needs to be addressed
+by the semantic conventions.
+
+### Instrumenting intermediaries
+
+Instrumenting intermediaries can be valuable for debugging configuration or
+performance issues, or for detecting specific intermediary failures.
+
+Stable semantic conventions for instrumenting intermediaries can be provided at
+a future point in time, but are not in scope for this OTEP.
+
+### Metrics
+
+Messaging semantic conventions for tracing and for metrics overlap and should
+be as consistent as possible. However, semantic conventions for metrics will be
+handled separately and are not in scope for this OTEP.
+
+### In-memory queues or channels
+
+Messaging semantic conventions are not meant for instrumenting in-memory queues
```

Comment:
@pyohannes I don't really agree with the statement "are not meant for instrumenting in-memory".

I'd prefer we said it can be used for in-memory, but explicitly catering to in-memory is out of scope for the otep.

The way it's worded at present gives the impression that it would be wrong to use them, which I don't think is the case

Reply:
@kenfinnigan I re-worded this section, trying to emphasize that everything falling in one of the scenarios laid out in the previous section should be supported (no matter whether the intermediary is a separate process or not).

Reply:
@pyohannes I'm good with it mentioning alignment with the scenarios, but saying in-memory channel is explicitly excluded from the conventions is a problem in my view.

With Quarkus, our use of an in-memory channel very much aligns with the scenarios outlined. The only difference is the intermediary is framework code and not a broker, but explicitly excluding in-memory channel indicates that what Quarkus does shouldn't follow the conventions.

I'd like the conventions to be applicable to the Quarkus use case as I would prefer not to have different conventions for reactive messaging in Quarkus between situations where AMQP or Kafka is the intermediary, vs the intermediary being an in-memory channel.

Maybe the wording needs to explicitly allow a reactive stream consisting of publishers and subscribers as being acceptable? That's the Quarkus use case for in-memory. Does the wording help at all?

Reply:
I removed the example mentioning the in-memory queue. I'm not concerned about the conventions not being applicable to Quarkus. I definitely see it more in the camp of full-fledged messaging systems than as a pattern of low-level asynchronous message passing. And I think the scenarios definitely apply to it.